tests/versioning_3_test.go

```diff
@@ -205,13 +210,129 @@ func (s *Versioning3Suite) testWorkflowWithPinnedOverride(sticky bool) {
 	s.verifyWorkflowVersioning(tv, vbUnpinned, tv.Deployment(), override, nil)
 
 	s.pollWftAndHandle(tv, sticky, nil,
-		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondWorkflowTaskCompletedRequest, error) {
+		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*taskpoller.TaskCompletedRequest, error) {
 			s.NotNil(task)
-			return respondCompleteWorkflow(tv, vbUnpinned), nil
+			return &taskpoller.TaskCompletedRequest{WorkflowTaskCompletedRequest: respondCompleteWorkflow(tv, vbUnpinned)}, nil
 		})
 	s.verifyWorkflowVersioning(tv, vbUnpinned, tv.Deployment(), override, nil)
 }
 
+func (s *Versioning3Suite) TestQueryWithPinnedOverride_NoSticky() {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testQueryWithPinnedOverride(false)
+		},
+	)
+}
+
+func (s *Versioning3Suite) TestQueryWithPinnedOverride_Sticky() {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testQueryWithPinnedOverride(true)
+		},
+	)
+}
+
+func (s *Versioning3Suite) testQueryWithPinnedOverride(sticky bool) {
+	tv := testvars.New(s)
+
+	if sticky {
+		s.warmUpSticky(tv)
+	}
+
+	wftCompleted := make(chan interface{})
+	s.pollWftAndHandle(tv, false, wftCompleted,
+		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*taskpoller.TaskCompletedRequest, error) {
+			s.NotNil(task)
+			return &taskpoller.TaskCompletedRequest{WorkflowTaskCompletedRequest: respondEmptyWft(tv, sticky, vbUnpinned)}, nil
+		})
+
+	override := makePinnedOverride(tv.Deployment())
+	we := s.startWorkflow(tv, override)
+
+	<-wftCompleted
+	s.verifyWorkflowVersioning(tv, vbUnpinned, tv.Deployment(), override, nil)
+	if sticky {
+		s.verifyWorkflowStickyQueue(we, tv.StickyTaskQueue())
+	}
+
+	s.pollAndQueryWorkflow(tv, sticky)
```

Comment:
Note: Here (after we poll and process a query), I don't think we should be adding extra lines of code verifying the workflow execution's versioning. I think this is because query tasks don't add an extra event in history and neither does the worker conduct a `RespondWorkflowTaskCompleted`, in which the execution updates the deployment attribute. 

Thus, a point to be noted is that even though a query is always redirected to it's latest deployment, the `workflowexecution.versioning_info.deployment` attribute will not be updated until a workflow task ends up in adding a workflow task completed event.

Reply:
[ my understanding might be a bit skewed above so feel free to correct me if I was wrong! ]